rust/agama-server/src/software/license.rs

```diff
@@ -0,0 +1,264 @@
+use std::{
+    fmt::Display,
+    fs::read_dir,
+    path::{Path, PathBuf},
+};
+use thiserror::Error;
+
+/// Represents a product license.
+///
+/// It contains the license ID and the list of languages that with a translation.
+#[serde_as]
+#[derive(Clone, Debug, Serialize)]
+pub struct License {
+    pub id: String,
+    #[serde_as(as = "Vec<DisplayFromStr>")]
+    pub languages: Vec<LanguageTag>,
+}
+
+/// Represents a license content.
+///
+/// It contains the license ID and the body.
+///
+/// TODO: in the future it might contain a title, extracted from the text.
+#[derive(Clone, Debug, Serialize)]
+pub struct LicenseContent {
+    pub id: String,
+    pub body: String,
+}
+
+/// Represents a repository of software licenses.
+///
+/// The repository consists of a directory in the file system which contains the licenses in
+/// different languages.
+///
+/// Each license is stored on a separate directory (e.g., "/usr/share/agama/eula/license.beta").
+/// The license diectory contains the default text (license.txt) and a set of translations (e.g.,
+/// "license.es.txt", "license.zh_CH.txt", etc.).
+#[derive(Clone)]
+pub struct LicensesRepo {
+    /// Repository path.
+    pub path: std::path::PathBuf,
+    /// Licenses in the repository.
+    pub licenses: Vec<License>,
+}
+
+impl LicensesRepo {
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        Self {
+            path: path.as_ref().to_owned(),
+            licenses: vec![],
+        }
+    }
+
+    /// Reads the licenses from the repository.
+    pub fn read(&mut self) -> Result<(), std::io::Error> {
+        let entries = read_dir(self.path.as_path())?;
+
+        for entry in entries {
+            let entry = entry?;
+            if entry.file_type()?.is_dir() {
+                let Ok(id) = entry.file_name().into_string() else {
+                    continue;
+                };
+                let license = License {
+                    id,
+                    languages: Self::find_translations(&entry.path())?,
+                };
+                self.licenses.push(license);
+            }
+        }
+        Ok(())
+    }
+
+    /// Finds a license with the given ID and language.
+    ///
+    /// If a translation is not found for the given language, it returns the default text.
+    pub fn find(&self, id: &str, language: &LanguageTag) -> Option<LicenseContent> {
+        let mut candidates: Vec<String> = vec![];
+        if let Some(territory) = &language.territory {
+            candidates.push(format!("license.{}_{}.txt", language.language, territory));
+        }
+        candidates.push(format!("license.{}.txt", language.language));
+        candidates.push("license.txt".to_string());
+
+        let license_path = candidates
+            .into_iter()
+            .map(|p| self.path.join(id).join(p))
+            .find(|p| p.exists())?;
+
+        let body: String = std::fs::read_to_string(license_path).ok()?;
+
+        Some(LicenseContent {
```

Comment:
I would like to see here at least a bit of logging, like candidates and found licese so we know from logs why given license is shown.

Reply:
IMHO that's too much logging just for retrieving a license (the logic is pretty straightforward). But I can live with that :-)